Add ability to specify source with each dependency

efirestone opened the pull request:
This adds the ability to add a `:source` flag to pod dependencies in a Podfile. This behavior mirrors in [Gemfiles](http://bundler.io/gemfile.html), allowing a dependency to be specified as such:

`pod 'JSONKit', '1.4', :source => 'http://github.com/MyOrg/Specs.git'`

Depends on https://github.com/CocoaPods/Core/pull/278

efirestone commented:
I didn't see a specific existing issue regarding this enhancement. There's some discussion [here](https://github.com/CocoaPods/CocoaPods/issues/3631), and I can file a new issue if you want one.

segiddins commented:
No need to file an issue

segiddins commented:
How does this handle transitive dependencies?

jafara commented:
:+1: this is a great change!

efirestone commented:
@segiddins Transitive dependencies are handled like so:
 - If a dependency is explicitly specified in the Podfile, then the :source repo will be used for that dependency if it comes up as a transitive dependency
 - If a dependency is not explicitly specified anywhere in the Podfile, then the normal file source rules apply (unchanged from before)

The simplest way to summarize is that it has the exact same behavior as a podspec given using :podspec or :path. Podfile dependencies get resolved first and are cached and reused for all subsequent requirements.


One small thing I noticed: as mentioned, the resolver relies on caching to make sure dependency resolutions are consistent. This works fine when calling `resolve`, but it means that if you call `search_for` directly before calling `resolve` then the wrong thing might be cached. This is true with the existing code for external sources (:path) as well.

To illustrate:
1. Have a Podfile with the dependency `pod 'MyPod', :path => '/path/to/1.0/MyPod.podspec'`
2. Use that podfile as a constructor argument in creating a `Resolver`
3. Before calling resolve, call `search_for(Dependency.new('MyPod', '1.0'))`

I would expect the resolver to give me back the Spec from the local path, but instead it will actually use the sources passed to the resolver constructor instead. If you call `resolve` first then the dependencies from the podfile will get passed to `search_for` first and will be the ones that are cached. The new :source flag behaves the same way.

Again, I think this is all fine, but I figured I'd call it out. Root cause here seems to be that `search_for` should be private so that you can't abuse it in this way but in reality has to be public to conform to `Molinillo::SpecificationProvider`.

efirestone commented:
Also, the code has a dependency on https://github.com/CocoaPods/Core/pull/278, so the tests should go green once we can point them at that. It looks like everything is pegged to HEAD, so is the standard development flow to merge the other PR first?

segiddins commented:
You'll need to point the Gemfile in this repo to your Core PR so the tests here can pass.
And I'm still not seeing tests for what will happen to transitive dependencies of dependencies with a custom source set -- one could argue that they ought to use the source for their parent dependency, though you then get into conflict situations... Either way, that scenario needs to be sufficiently tested.

segiddins commented on lib/cocoapods/sources_manager.rb:
```diff
@@ -418,6 +438,24 @@ def master_repo_functional?
+        @aggregates_by_repos ||= {}
+
+        # Use the cached version if it exists.
+        if result = @aggregates_by_repos[repos]
```
all this can be replaced with `@aggregates_by_repos[repos] ||= Source::Aggregate.new(repos)`

efirestone replied:
Fixed

segiddins commented on spec/unit/installer/analyzer_spec.rb:
```diff
@@ -88,6 +101,28 @@ module Pod
+        source.stubs(:repo).returns('/repo/cache/path')
+        SourcesManager.expects(:find_or_create_source_with_url).with(repo_url).returns(source)
+        SourcesManager.stubs(:git_repo?).with(source.repo).returns(true)
+        SourcesManager.expects(:update).once.with(source.name).returns
```
what should this be returning?

efirestone replied:
Fixed (it returns void, so that last call is unnecessary)

efirestone commented on spec/unit/installer/analyzer_spec.rb:
```diff
@@ -62,6 +62,19 @@ module Pod
         @analyzer.update_repositories
       end
 
+      it 'does not update sources if there are no dependencies' do
```
This is a small optimization that came along with my changes. If a Podfile has no dependencies specified, then it won't bother updating any of the sources. This seems safe to me, but I'd like confirmation from the veterans on this.

efirestone commented on spec/unit/installer/analyzer_spec.rb:
```diff
@@ -474,6 +509,7 @@ module Pod
             it 'raises if no specs repo with that URL could be added' do
               podfile = Podfile.new do
                 source 'not-a-git-repo'
+                pod 'JSONKit', '1.4'
```
Because of the change mentioned above, these Podfiles needed a dependency in them in order to actually update their source.

efirestone commented on spec/unit/resolver_spec.rb:
```diff
@@ -441,46 +441,6 @@ module Pod
         e.message.should.match(/`pod update CocoaLumberjack`/)
       end
 
-      it 'consults all sources when finding a matching spec' do
```
These moved down into a `describe 'Multiple Sources'` section.

efirestone commented on spec/unit/resolver_spec.rb:
```diff
@@ -550,6 +510,152 @@ def resolve
+        spec.defined_in_file.should == fixture('spec-repos/master/Specs/JSONKit/1.5pre/JSONKit.podspec.json')
+      end
+
+      it 'uses explicit source repos for a dependency even when it\'s transitive' do
```
Tests that if a dependency has transitive dependencies, and those transitive dependencies are also top-level dependencies, then the transitive dependency is resolved using the same resolution as the top-level dependency.

efirestone commented on spec/unit/resolver_spec.rb:
```diff
@@ -550,6 +510,152 @@ def resolve
+        possible_specs.first.defined_in_file.should == fixture('spec-repos/test_repo/JSONKit/1.4/JSONKit.podspec')
+      end
+
+      it 'uses global source repos for resolving a transitive dependency even ' \
```
Tests that for a top-level dependency with an explicit source, its transitive dependencies can be resolved by any global sources in the Podfile (unless the transitive dependency is explicitly pegged to a source or path, as tested above).

This is the same behavior that Gemfiles use, and also the one that makes the most sense to me. In most cases I think a dependency will be specified with an explicit source because it's a company internal pod. That internal pod should still be able to depend on public pods (like JSONKit or AFNetworking) without needing to clone those public pods into the internal podspec repo. If the internal pod has dependencies on other internal pods then the internal podspec source can be included at the global level as well (preferred), or those dependencies can be enumerated explicitly at the Podfile level with :source flags.

efirestone commented:
Added some transitive dependency tests. See other comments for discussion on the behavior. Also updated the Gemfile and addressed your comments. Thanks for looking it over @segiddins!

efirestone commented:
This is ready for a review. Much appreciated!

segiddins commented on CHANGELOG.md:
```diff
@@ -17,6 +17,11 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
+  [Eric Firestone](https://github.com/efirestone)
+  [#4486](https://github.com/CocoaPods/CocoaPods/pull/4486)
+
+* Improve `pod search` performance while using _`--full`_ flag  
```
this line shouldn't be there

segiddins commented on spec/unit/resolver_spec.rb:
```diff
@@ -550,6 +510,152 @@ def resolve
+          %w(AFNetworking AFNetworking/NSURLConnection AFNetworking/NSURLSession AFNetworking/Reachability) +
+            %w(AFNetworking/Security AFNetworking/Serialization AFNetworking/UIKit CrossRepoDependent)
+
+        afnetworking_spec = specs.find { |s| s && s.name == 'AFNetworking' }
```
the `s &&` bit shouldn't be necessary

segiddins commented on spec/unit/resolver_spec.rb:
```diff
@@ -550,6 +510,152 @@ def resolve
+            %w(AFNetworking/Security AFNetworking/Serialization AFNetworking/UIKit CrossRepoDependent)
+
+        afnetworking_spec = specs.find { |s| s && s.name == 'AFNetworking' }
+        afnetworking_spec.nil?.should == false
```
`afnetworking_spec.should.not.be.nil`

segiddins commented:
Generally looks good to me

efirestone commented:
Comments addressed. Thanks for looking it over.

efirestone commented:
Who are good candidates to take a second look at this? @orta? @neonichu?

manuyavuz commented on CHANGELOG.md:
```diff
@@ -20,6 +20,10 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#4487](https://github.com/CocoaPods/CocoaPods/pull/4487)
 
+* Add support for specifying :source with a pod dependency.  
```
I think your changelog entry should be at the top in order to preserve historical flow.

efirestone replied:
Done and done.

neonichu commented:
Looks good to me as well — thanks for the comprehensive comments, @efirestone :+1:

manuyavuz commented:
@efirestone, it just needs a rebase to be 🚀. Thank you for your effort !

manuyavuz commented:
And you should revert the Gemfile because changes in Core repo has been merged.

efirestone commented:
Should be good to go! Thanks all!

neonichu commented:
Awesome, thank you :+1:
